src/h5web/visualizations/index.tsx

```diff
@@ -1,101 +1,152 @@
-    Component: RawVis,
-    supportsDataset: () => true,
-    defaultMappingState: () => undefined,
+    Component: RawViscontainer,
+    supportsEntity: isDataset,
   },
 
   [Vis.Scalar]: {
     Icon: FiCode,
-    Component: ScalarVis,
-    supportsDataset: (dataset) => {
-      const { type, shape } = dataset;
-      return isBaseType(type) && isScalarShape(shape);
+    Component: ScalarVisContainer,
+    supportsEntity: (entity) => {
+      return (
+        isDataset(entity) &&
+        isBaseType(entity.type) &&
+        isScalarShape(entity.shape)
+      );
     },
-    defaultMappingState: () => undefined,
   },
 
   [Vis.Matrix]: {
     Icon: FiGrid,
-    Component: MappedMatrixVis,
-    supportsDataset: (dataset) => {
-      const { type, shape } = dataset;
-      return isBaseType(type) && isSimpleShape(shape) && shape.dims.length >= 1;
-    },
-    defaultMappingState: (dataset) => {
-      const { dims } = dataset.shape as HDF5SimpleShape;
-      return dims.length === 1
-        ? ['x']
-        : [...range(dims.length - 2).fill(0), 'y', 'x'];
+    Component: MatrixVisContainer,
+    supportsEntity: (entity) => {
+      return (
+        isDataset(entity) &&
+        isBaseType(entity.type) &&
+        isSimpleShape(entity.shape) &&
+        entity.shape.dims.length >= 1
+      );
     },
   },
 
   [Vis.Line]: {
     Icon: FiActivity,
     Toolbar: LineToolbar,
-    Component: MappedLineVis,
-    supportsDataset: (dataset) => {
-      const { type, shape } = dataset;
+    Component: LineVisContainer,
+    supportsEntity: (entity) => {
       return (
-        isNumericType(type) && isSimpleShape(shape) && shape.dims.length >= 1
+        isDataset(entity) &&
+        isNumericType(entity.type) &&
+        isSimpleShape(entity.shape) &&
+        entity.shape.dims.length >= 1
       );
     },
-    defaultMappingState: (dataset) => {
-      const { dims } = dataset.shape as HDF5SimpleShape;
-      return [...range(dims.length - 1).fill(0), 'x'];
-    },
   },
 
   [Vis.Heatmap]: {
     Icon: FiMap,
     Toolbar: HeatmapToolbar,
-    Component: MappedHeatmapVis,
-    supportsDataset: (dataset) => {
-      const { type, shape } = dataset;
+    Component: HeatmapVisContainer,
+    supportsEntity: (entity) => {
+      return (
+        isDataset(entity) &&
+        isNumericType(entity.type) &&
+        isSimpleShape(entity.shape) &&
+        entity.shape.dims.length >= 2
+      );
+    },
+  },
+
+  [Vis.NxSpectrum]: {
+    Icon: FiActivity,
+    Toolbar: LineToolbar,
+    Component: NxSpectrumContainer,
+    supportsEntity: (entity, metadata) => {
+      if (!isGroup(entity) || !isNxDataGroup(entity)) {
+        return false;
+      }
+
+      const dataset = getSignalDataset(entity, metadata);
+      if (!dataset) {
+        return false;
+      }
+
+      const interpretation = getAttributeValue(dataset, 'interpretation');
+      if (interpretation) {
+        return interpretation === NxInterpretation.Spectrum;
+      }
```

Comment:
Hmm problem... The error boundary is located in `Visualizer`, and so is the call to `useSupportedVis` which calls `supportsEntity`. If I throw in a `supportsEntity` function, it won't be caught by the boundary.

We could also throw if the NX data group doesn't have an `signal` attribute, or if the attribute points to a non-existing dataset... The HDF5 file could misuse NeXuS in many ways... To me, this is a different class of errors than the assertions that we have in the containers, which are programmer errors and have to break things (i.e. stop rendering the vis component and show an error).

The need here is for the creator of a file (or a person involved in the file's existence) to be able to debug what's wrong with it. Not everyone looking at a file would have this debugging need, so displaying an error in the same way as we display programmer and data fetching errrors might worsen the UX for these people.

What we could do is simply log errors to the console. Alternatively, we could throw errors from `supportsEntity`, catch them in `useSupportedVis` and return the list of errors that got caught to `Visualizer` so it can display them in a smart way (e.g. in a tooltip or an accordion).

Should we put this aside for this PR and talk about it in an issue?

Reply:
> Hmm problem... The error boundary is located in Visualizer, and so is the call to useSupportedVis which calls supportsEntity. If I throw in a supportsEntity function, it won't be caught by the boundary.

I feared as such.

> Should we put this aside for this PR and talk about it in an issue?

Yes, sounds good :+1:

Reply:
See #254